### PR TITLE
WT-13442 Do not allow out of order truncate commit timestamp

### PR DESCRIPTION
**Pull request by kommiharibabu**
WiredTiger doesn't allow out of order commit timestamp and it return an error whenever an operation is trying to commit with a timestamp less than the existing timestamp on an update. But it missed to verify the truncate commit timestamp.

The WiredTiger timestamp usage check is extended for the truncate operations also doesn't allow an out of order commit timestamp to be performed.

### CODE REVIEW

File: test/suite/test_truncate28.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import wiredtiger, wttest
+from helper import simulate_crash_restart
+
+# test_truncate28.py
+# Test that out of order truncate commit timestamp doesn't allowed.
```
Comment (Sean04) [inline test/suite/test_truncate28.py:33]:
```suggestion
# Test that out of order commit timestamps aren't allowed when performing truncate operations.
```

File: test/suite/test_truncate28.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        # Update stable timestamp and checkpoint.
+        stable_ts = (int) (nrows / 2)
+        self.conn.set_timestamp('stable_timestamp=' + self.timestamp_str(stable_ts))
+        self.session.checkpoint()
+
+        # Evict everything.
+        self.evict_cursor(uri, nrows + 1)
+        self.session.checkpoint()
+
+        # Do a fast truncate with a commit timestamp is less than the previous durable timestamp on the page.
```
Comment (Sean04) [inline test/suite/test_truncate28.py:90]:
```suggestion
        # Do a fast truncate with a commit timestamp that is less than the previous durable timestamp on the page.
```

File: src/include/btree.h [hunk 3]
```diff
@@ -93,6 +93,23 @@ typedef enum { /* Start position for eviction walk */
 /* An invalid btree file ID value. ID 0 is reserved for the metadata file. */
 #define WT_BTREE_ID_INVALID UINT32_MAX
 
+/*
+ * WT_ADDR_COPY --
+ *	We have to lock the WT_REF to look at a WT_ADDR: a structure we can use to quickly get a
+ * copy of the WT_REF address information.
+ */
+struct __wt_addr_copy {
```
Comment (Sean04) [inline src/include/btree.h:101]: Why did this get moved?

File: src/include/txn_inline.h [hunk 4]
```diff
@@ -569,7 +676,7 @@ __wt_txn_modify_page_delete(WT_SESSION_IMPL *session, WT_REF *ref)
     ref->page_del->txnid = txn->id;
 
     if (__txn_should_assign_timestamp(session, op))
-        __txn_op_delete_commit_apply_page_del_timestamp(session, op->u.ref);
+        __txn_op_delete_commit_apply_page_del_timestamp(session, op);
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/include/txn_inline.h]: Yes, fixed.

File: src/include/btree.h (hunk 3 above)
Comment (Siddhartha8899) [reply src/include/btree.h:101]: What about putting this struct is `bt_mem.h` instead? Below where `__wt_addr` is defined?

File: src/include/txn_inline.h [hunk 5]
```diff
@@ -471,7 +574,7 @@ __wt_txn_op_set_timestamp(WT_SESSION_IMPL *session, WT_TXN_OP *op)
     txn = session->txn;
 
     if (!__txn_should_assign_timestamp(session, op))
-        return;
+        return (0);
```
Comment (Siddhartha8899) [inline src/include/txn_inline.h:577]: Do we not need to check for timestamps in the if branch (`if (F_ISSET(txn, WT_TXN_PREPARE))`)?

Comment (Siddhartha8899) [reply src/include/txn_inline.h:577]: Ops sorry my bad, we only check for commit.

File: src/include/txn_inline.h [hunk 6]
```diff
@@ -363,34 +363,54 @@ __wt_txn_op_delete_apply_prepare_state(WT_SESSION_IMPL *session, WT_REF *ref, bo
...
     page_del = ref->page_del;
 
     WT_ASSERT(session, WT_REF_GET_STATE(ref) == WT_REF_LOCKED);
 
     if (page_del != NULL && page_del->timestamp == WT_TS_NONE) {
+
+        /* Validate the commit timestamp against the maximum durable timestamp on the page. */
+        WT_ENTER_GENERATION(session, WT_GEN_SPLIT);
+        WT_WITH_BTREE(session, op->btree, addr_found = __wt_ref_addr_copy(session, ref, &addr));
+        if (addr_found)
```
Comment (Siddhartha8899) [inline src/include/txn_inline.h:387]: Shouldn't this be an assert, as a deleted page should always have an addr?

File: src/include/txn_inline.h [hunk 7]
```diff
@@ -363,34 +363,54 @@ __wt_txn_op_delete_apply_prepare_state(WT_SESSION_IMPL *session, WT_REF *ref, bo
...
+            ret = __wt_txn_timestamp_usage_check(session, op, txn->commit_timestamp,
+              WT_MAX(addr.ta.newest_start_durable_ts, addr.ta.newest_stop_durable_ts));
+        WT_LEAVE_GENERATION(session, WT_GEN_SPLIT);
```
Comment (Siddhartha8899) [inline src/include/txn_inline.h:390]: Shouldn't we leave the generation sooner (before checking the timestamps)?

File: src/include/btree.h (hunk 3 above)
Comment (kommiharibabu) [reply src/include/btree.h:101]: My first change was in the `bt_mem.h` file only, but it has a dependency with a macro from the `btree.h`. Due to that reason, I moved it into the `btree.h` file.

File: src/include/txn_inline.h (hunk 7 above)
Comment (kommiharibabu) [reply src/include/txn_inline.h:390]: As soon as we leave the generation, the `addr.addr` can be freed. But anyway we are not using that memory here, so the generation can be released sooner when we get the copy, but not releasing it sooner also doesn't cause any problems other than holding the memory for some more time if it is freed in parallel.

**Review by Sean04 — APPROVED** [review]
LGTM

File: src/include/btree.h (hunk 3 above)
Comment (Siddhartha8899) [reply src/include/btree.h:101]: I see, perhaps it's nit pick and extra work but wanted to point it out any way - 
It seems reasonable renaming this macro to `WT_ADDR_MAX_COOKIE`; and moving it along with `__wt_addr_copy` to `btmem.h` - for the sake of similar structs being in the same file. 

I leave that up to you tho, so feel free to resolve this one :-)

File: src/include/txn_inline.h (hunk 7 above)
Comment (Siddhartha8899) [reply src/include/txn_inline.h:390]: great makes sense to leave it this way, thanks!

File: src/include/btree.h (hunk 3 above)
Comment (kommiharibabu) [reply src/include/btree.h]: Yes, it is a nice idea. I moved the macro and also the structure into the btmem.h file.

File: src/include/txn_inline.h (hunk 6 above)
Comment (kommiharibabu) [reply src/include/txn_inline.h]: Sometimes a deleted page may not have an addr. But in this scenario, as it is a fast truncated page, it must have an address. I changed the `if` check into an assert_always.

**Review by Siddhartha8899 — APPROVED** [review]

**Pull request merged**